helix-term/src/application.rs

```diff
@@ -144,22 +155,63 @@ impl Application {
 
             use futures_util::StreamExt;
 
+            #[cfg(not(windows))]
             tokio::select! {
                 biased;
 
                 event = reader.next() => {
                     self.handle_terminal_events(event)
                 }
+                Some(signal) = self.signals.next() => {
+                    use helix_view::graphics::Rect;
+                    match signal {
+                        signal::SIGTSTP => {
+                            self.compositor.save_cursor();
+                            self.restore_term().unwrap();
+                            low_level::emulate_default_handler(signal::SIGTSTP).unwrap();
+                        }
+                        signal::SIGCONT => {
+                            self.claim_term().await.unwrap();
+                            // redraw the terminal
+                            let Rect { width, height, .. } = self.compositor.size();
+                            self.compositor.resize(width, height);
+                            self.compositor.load_cursor();
+                            self.render();
+                        }
+                        _ => unreachable!(),
+                    }
+                }
                 Some((id, call)) = self.editor.language_servers.incoming.next() => {
                     self.handle_language_server_message(call, id).await;
                     // limit render calls for fast language server messages
                     let last = self.editor.language_servers.incoming.is_empty();
                     if last || last_render.elapsed() > deadline {
                         self.render();
                         last_render = Instant::now();
-                    }
+                    }                }                Some(callback) = self.jobs.futures.next() => {
+                    self.jobs.handle_callback(&mut self.editor, &mut self.compositor, callback);
+                    self.render();
                 }
-                Some(callback) = self.jobs.futures.next() => {
+                Some(callback) = self.jobs.wait_futures.next() => {
+                    self.jobs.handle_callback(&mut self.editor, &mut self.compositor, callback);
+                    self.render();
+                }
+            }
+            #[cfg(windows)]
```

Comment:
Can't we do this on the specific branch in the loop above? Maintaining two loops here is not ideal

Reply:
No, tokio doesn't support that. I am trying to add that and it's hard. https://github.com/tokio-rs/tokio/pull/3977 Currently out of ideas but someone suggested some ideas which I can try out.